### ISSUE TITLE
OMG Network Upgrade to deploy v0.2

### ISSUE THREAD

**Issue by kasima** [omisego/elixir-omg#637]
We need to perform a "hard" network upgrade to deploy a new version of plasma-contracts and the features of elixir-omg that are dependent of the new version.

This hard upgrade will require all token holders on Ari to perform an exit and re-deposit into the new chain.

**Comment by Pongch** [comment]
There are multiple related steps/tasks/parts in the upgrade story. Currently:
1. [Tests for Non-Consolidated Exits ](https://github.com/omisego/elixir-omg/issues/615) is happening this week
2. Tests for Consolidated Exits (TODO)
3. Actual Coordination of Exit and re-deposit process (TODO)

**Comment by Pongch** [comment]
I will make 2. and 3. as individual GH issues and we will have the discussions there

**Comment by Pongch** [comment]
There are some current blockers to upgrade path:

- We need a quick solution to handling process exits/ giving visibility now before the upgrade happens. Otherwise the Exit queue will be clogged with transactions and plasma-dog/ ODP users will not have any oversight on whether they received their funds back on the Rootchain. 

- V0.2 Compatibility needs to happen on a) omg-js, js-starter-kit before we can spin up a V0.2 cluster. Otherwise people will not be able to interact with the network

**Comment by Pongch** [comment]
RE: Temporary Solution to dealing with clearing out the clogged exit queue, we do have a few options, given the time frame we have CC @pdobacz :
1/ User calling `processExit` themselves. This is the default way provided by upgrade scripts.
pros: 
-  user dealing with their own funds
cons:
- user still does not know how many exits there are, or how many is in the queue, or where their exits are.

2/ We (OmiseGO) constantly poll `ExitQueue.currentSize()` for number of exits and calling `processExit` for users.
pros:
- 1 less upgrade step for user
cons:
- require building a service ourselves (albeit a temporary one)
- require a lot of gas

IMO, point 2 is an option we can take

**Comment by pdobacz** [comment]
:+1: 2/ sounds good, also there is a twist we can have - we only monitor `ExitQueue.currentSize()` and postpone work on auto-process exiting for the users until we see it bloats. Just to cover the scenario when few users decide to exit.

In case only few users decide to exit and the queue doesn't get clogged up, it's better to have the pro "user dealing with their own funds" for exploration of the UX and a more real-life experience there.

An unrelated question - how are you figuring out which `ExitQueues` to poll? there's one per each token `addToken`ed. Are we leveraging the `fee_specs` from the child chain deployment?

Oh, one more: did we `addToken` all the tokens people will consider exiting?

**Comment by Pongch** [comment]
> In case only few users decide to exit and the queue doesn't get clogged up, it's better to have the pro "user dealing with their own funds" for exploration of the UX and a more real-life experience there.> 

Yup, great idea ! 

>An unrelated question - how are you figuring out which ExitQueues to poll? there's one per each token addTokened. Are we leveraging the fee_specs from the child chain deployment?>

Yup, from fee_specs and Plasma Dog tokens. Another alternative is to get the logs of AddToken() events.

>Oh, one more: did we addToken all the tokens people will consider exiting?>

I might have missed this completely in ERC20 tokens 😨 I don't think even Hoard is aware that this steps needs to be taken first, but I can check with Radek. I will make this crystal clear in the Exit + Upgrade guides.

Luckily enough for us, most people have been transacting with ETH on Ari. So aside from edge-cases and Plasma Dog, people will be exiting ETH mostly

**Comment by pdobacz** [comment]
> Luckily enough for us, most people have been transacting with ETH on Ari. So aside from edge-cases and Plasma Dog, people will be exiting ETH mostly

Not sure what CP did and whether their tokens are in the `fee_specs` (unless they're planning to exit, possibly not since this is not within the scope of their usecase).

Other than that, I think such exotic ERC20 tokens won't have their queues clogged up. You're right -  (possibly) facilitating the "managed" exits done by Plasma Dog a bit + explanation in guides should do.
